### PR TITLE
libxnd, libndtypes: refactor add support for darwin

### PR DESCRIPTION
**Pull request by costrouc**
###### Motivation for this change

Prior build instructions were hacks because I did not understand that the configure script was choosing `ld` as the linker instead of using `gcc` to properly link. This commit fixes these issues and simplifies the build. 

###### Things done

libxnd: refactor
libndtypes: refactor

Please ensure that this build works on darwin.

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [X] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by costrouc** [comment]
For some reason the `configure` script by default chooses `ld` as the linker. This causes errors because the `Makefile` assumes that `gcc` or `clang` is used for the linker. Is there a more elegant way to do this?

**Comment by LnL7** [comment]
Yes that's very normal, gcc and clang are not linkers. Could you explain or post the error?

**Comment by costrouc** [comment]
I am aware that gcc and clang are not linkers (they are wrappers around preprocessing, compiling, and linking). From what I know they can be used as linkers and call `ld` with appropriate arguments. Here is the error if I do not manually set `cc` as the linker. 

```
gcc -I.. -Wall -Wextra -std=c11 -pedantic -O2 -g -fPIC -c seq.c -o .objs/seq.o
gcc -I.. -Wall -Wextra -std=c11 -pedantic -O2 -g -fPIC -c symtable.c -o .objs/symtable.o
gcc -I.. -Wall -Wextra -std=c11 -pedantic -O2 -g -fPIC -c util.c -o .objs/util.o
gcc -I.. -Wall -Wextra -std=c11 -pedantic -O2 -g -fPIC -c values.c -o .objs/values.o
ld -shared -Wl,-soname,libndtypes.so.0 -o libndtypes.so.0.2.0dev3 .objs/alloc.o .objs/attr.o .objs/context.o .objs/copy.o .objs/encodings.o .objs/equal.o .objs/grammar.o .objs/io.o .objs/lexer.o .objs/match.o .objs/ndtypes.o .objs/parsefuncs.o .objs/parser.o .objs/seq.o .objs/symtable.o .objs/util.o .objs/values.o compat/.objs/bpgrammar.o compat/.objs/bplexer.o compat/.objs/import.o
/nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/ld: unrecognized option '-Wl,-soname,libndtypes.so.0'
/nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/ld: use the --help option for usage information
make[1]: *** [Makefile:57: libndtypes.so.0.2.0dev3] Error 1
make[1]: Leaving directory '/build/source/libndtypes'
make: *** [Makefile:23: default] Error 2
builder for '/nix/store/5gplhcy8rxmiabvw43agnwpkflrvz29l-libndtypes-0.2.0dev3.drv' failed with exit code 2
error: build of '/nix/store/5gplhcy8rxmiabvw43agnwpkflrvz29l-libndtypes-0.2.0dev3.drv' failed
```

`'-Wl,-soname,libndtypes.so.0'` is a argument that is passed to `gcc` or `clang` that is appropriately passed to `ld`.

**Comment by LnL7** [comment]
Ah I see, the build is passing compiler specific flags to ld like I thought. `-Wl` is a compiler flag to specify extra flags to pass to the linker. Could you rephrase the comment to avoid confusion.

That should be, so using `LD=cc` basically turns the incorrect usage into the first variant.

```sh
$CC -Wl,-soname,libndtypes.so.0 ...
# or
$LD -soname libndtypes.so.0 ...
```

**Comment by costrouc** [comment]
Ok I have updated the pull request with a better comment for why cc was used as a linker. Its ready to merge.

I still am not convinced this is incorrect usage. While in this case using ld would have been simple enough. For some complex libraries this is not always the case. Take for example `mpicc`, `mpif90` these are wrappers around gcc with linker and special compilation flags. So similarly using `gcc` for linking simplifies the process by not having to worry about the location of glibc, lm, ld-linux-x86-64.so.2, etc. Using `-Wl,--verbose` you can print all of the work done by `gcc` for linking.

**Comment by LnL7** [comment]
I only have a vague understanding of this, somebody like @Ericson2314 can probably explain this better. But we explicitly set `LD=ld` in the stdenv which is probably what exposed the issue.

File: pkgs/development/libraries/libndtypes/default.nix [hunk 1]
```diff
@@ -14,12 +14,14 @@ stdenv.mkDerivation rec {
     sha256 = "0dpvv13mrid8l5zkjlz18qvirz3nr0v98agx9bcvkqbiahlfgjli";
   };
 
-  makeFlags = [ "CONFIGURE_LDFLAGS='-shared'" ];
+  # Override linker with cc (symlink to either gcc or clang)
+  # Library expects to use cc for linking
+  configureFlags = [ "LD=cc" ];
```
Comment (Ericson2314) [inline pkgs/development/libraries/libndtypes/default.nix]: `${stdenv.cc.targetPrefix}cc` instead of `cc`, for sake of cross

File: pkgs/development/libraries/libxnd/default.nix [hunk 2]
```diff
@@ -17,10 +17,9 @@ stdenv.mkDerivation rec {
 
   buildInputs = [ libndtypes ];
 
-  configureFlags = [ "XND_INCLUDE='-I${libndtypes}/include'"
-                     "XND_LINK='-L${libndtypes}/lib'" ];
-
-  makeFlags = [ "CONFIGURE_LDFLAGS='-shared'" ];
+  # Override linker with cc (symlink to either gcc or clang)
+  # Library expects to use cc for linking
+  configureFlags = [ "LD=cc" ];
```
Comment (Ericson2314) [inline pkgs/development/libraries/libxnd/default.nix]: ditto

**Comment by Ericson2314** [comment]
Yeah it is done for cross on other things to force configure scripts to use the right linker. But then some configure scripts are crazy and have *defaults* like `LD ?= gcc` for this. I've seen a `CCLD` environment variable instead, which is a much better way to use a C compiler for linking but also be independent of`CC`. That is what upstream ought to use.

**Comment by Ericson2314** [comment]
@costrouc I'd change "compiler to determine the linker" to "using a compiler to do the linking". The compiler will modify how you link things in subtle ways to pick up the C/C++ runtime things, etc.

**Comment by costrouc** [comment]
Should be ready for a build now
